### PR TITLE
ACLK connection and protocol improvements

### PR DESCRIPTION
**Pull request by stelfrag**
Improve the ACLK startup sequence using popcorn-ing. The agent will enter a stable state before sending update events to the cloud. Enhance the payload with extra fields (version, timestamp) . Detect new (or deleted) collectors and send the appropriate message to the cloud.

### CODE REVIEW

File: aclk/agent_cloud_link.c [hunk 1]
```diff
@@ -133,6 +141,41 @@ struct aclk_query_queue {
...
+    static  int fail = -1;
+    unsigned long int delay;
+
+    if (!mode || fail == -1) {
+        srandom(time(NULL));
+        fail = mode-1;
+        return 0;
+    }
+
+    delay = (1 << fail);
```
Comment (thiagoftsm) [inline aclk/agent_cloud_link.c]: I am studying the code from top to bottom, so it is possible that my answer is below, but case it is not, I will do a question. Is there any possibility of `fail` to be bigger than 64? Case it has this possibility, we need to bring an initial test here to avoid overflow.

**Comment by underhood** [comment]
question for the future: when malformed request sent it is logged properly in error log. Should we also inform cloud that we didn't understand?

File: web/api/formatters/rrdset2json.c [hunk 2]
```diff
@@ -25,30 +25,45 @@ void rrdset2json(RRDSET *st, BUFFER *wb, size_t *dimensions_count, size_t *memor
...
+        );
+
+    buffer_sprintf(wb,
+                "\t\t\t\"first_entry\": %ld,\n"
+        , first_entry_t //rrdset_first_entry_t(st)
+    );
+
+    if (likely(!skip_volatile))
+        buffer_sprintf(wb,
+
```
Comment (thiagoftsm) [inline web/api/formatters/rrdset2json.c]: I think we can remove this blank line.

**Comment by thiagoftsm** [comment]
> question for the future: when malformed request sent it is logged properly in error log. Should we also inform cloud that we didn't understand?

I think this would be a good idea. And I would extend your question, is it possible to classify the errors and send a specific sample of them to the cloud?

File: aclk/agent_cloud_link.c [hunk 3]
```diff
@@ -173,20 +222,26 @@ struct aclk_query *aclk_query_find_position(time_t time_to_run)
 }
 
 // Need to have a lock before calling this
-struct aclk_query *aclk_query_find(char *topic, char *data, char *msg_id, char *query)
+struct aclk_query *aclk_query_find(char *topic, char *data, char *msg_id, char *query, ACLK_CMD cmd, struct aclk_query **last_query)
```
Comment (amoss) [inline aclk/agent_cloud_link.c]: Every function in this file that is not called from elsewhere should be static.

File: aclk/agent_cloud_link.c [hunk 4]
```diff
@@ -173,20 +222,26 @@ struct aclk_query *aclk_query_find_position(time_t time_to_run)
 }
 
 // Need to have a lock before calling this
```
Comment (amoss) [inline aclk/agent_cloud_link.c]: Be more explicit - must have the query lock.

File: aclk/agent_cloud_link.c [hunk 5]
```diff
@@ -196,31 +251,50 @@ struct aclk_query *aclk_query_find(char *topic, char *data, char *msg_id, char *
...
+    }
+
     run_after = now_realtime_sec() + run_after;
 
     QUERY_LOCK;
-    tmp_query = aclk_query_find(topic, data, msg_id, query);
+    struct aclk_query *last_query = NULL;
+
+    //last_query = NULL;
+    tmp_query = aclk_query_find(topic, data, msg_id, query, aclk_cmd, &last_query);
```
Comment (amoss) [inline aclk/agent_cloud_link.c]: Removing duplicate queries was not in the spec - what if the far end is trying to reduce latency by pipelining, i.e. issuing multiple queries in sequence so that the req/responses overlap?

File: aclk/agent_cloud_link.c [hunk 6]
```diff
@@ -303,7 +359,27 @@ struct aclk_query *aclk_queue_pop()
 
     this_query = aclk_queue.aclk_query_head;
 
-    if (this_query->run_after > now_realtime_sec()) {
+    // Get rid of the deleted entries
```
Comment (amoss) [inline aclk/agent_cloud_link.c]: Deletion in a queue is fast - why are you doing deferred deletion instead?

File: aclk/mqtt.c [hunk 7]
```diff
@@ -113,10 +79,20 @@ int lws_wss_client_initialized = 0;
 int mqtt_over_websockets = 1;
```
[the remark this answers is not in the document]
Comment (underhood) [reply aclk/mqtt.c]: It might prove usefull when we have MQTT over HTTPS etc. etc.

File: aclk/mqtt.c [hunk 8]
```diff
@@ -235,7 +210,7 @@ int _mqtt_lib_init(char *aclk_hostname, int aclk_port, void (*on_connect)(void *
 #endif
 
     if(!mqtt_over_websockets) {
```
[the remark this answers is not in the document]
Comment (underhood) [reply aclk/mqtt.c]: Hi as i mention before in future we plan MQTT over HTTPS.

File: aclk/agent_cloud_link.c [hunk 9]
```diff
@@ -40,11 +50,11 @@ int cloud_to_agent_parse(JSON_ENTRY *e)
...
-                data->topic = strdupz(e->data.string);
+                data->callback_topic = strdupz(e->data.string);
                 break;
             }
             if (!strcmp(e->name, ACLK_JSON_IN_URL)) {
-                data->url = strdupz(e->data.string);
+                data->payload = strdupz(e->data.string);
                 break;
             }
             break;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: No reason for atol just copy paste from other part of the agent code. 

Actually there is a bug here. When the field is number the  `data.string` is not available. Other parts of the code use original_string. So it will change to atoi along with that bug fix.

File: aclk/agent_cloud_link.c [hunk 10]
```diff
@@ -5,20 +5,30 @@
...
+
+    uuid_generate(uuid);
+    uuid_unparse(uuid, uuid_str);
+
+    return uuid_str;
+}
+
 int cloud_to_agent_parse(JSON_ENTRY *e)
 {
     struct aclk_request *data = e->callback_data;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: Good catch!

File: aclk/mqtt.c [hunk 11]
```diff
@@ -71,38 +65,10 @@ inline const char *_link_strerror(int rc)
...
+    UNUSED(mosq);
+    UNUSED(obj);
 
     aclk_handle_cloud_request(msg->payload);
-
-    //info("Received type=[%s], msg-id=[%s], topic=[%s], url=[%s]",cloud_to_agent.type_id, cloud_to_agent.msg_id, cloud_to_agent.topic, cloud_to_agent.url);
-
 }
 
 int lws_wss_client_initialized = 0;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: True 
Although the reconnect is not in place the `aclk_init` function already has an action (to allow for reinit) which should propagate into `_link_lib_init` that would reset this variable internally.

File: aclk/agent_cloud_link.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (thiagoftsm) [reply aclk/agent_cloud_link.c]: Considering that there was not a specific motive, I will give suggestion, `atoi`, `atol` and `atoll` are functions that call `strtol`, case it is possible, call direct this last casting the result ;).

File: aclk/agent_cloud_link.c (hunk 6 above)
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: There is code to delete in place. Needs some optimization. Will be done in another issue along with the deferred deletion code (and cleanup of the data structure)

File: aclk/agent_cloud_link.c (hunk 5 above)
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: Will remove / discuss in future issue

File: aclk/agent_cloud_link.c (hunk 3 above)
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: Will add in future issue

File: aclk/mqtt.c [hunk 12]
```diff
@@ -7,34 +7,33 @@
...
 #ifndef ENABLE_ACLK
 
 inline const char *_link_strerror(int rc)
 {
-    (void) rc;
+    UNUSED(rc);
     return "no error";
 }
 
 int _link_event_loop(int timeout)
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: No reason actually, my way of indicating "very" internal to this part of the code :smile:

File: aclk/mqtt.c [hunk 13]
```diff
@@ -359,7 +334,7 @@ int _link_event_loop(int timeout)
 {
     if(mqtt_over_websockets)
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: This (among) other things needs to be done with the integration of the disconnect callback. We will need to cleanup various pieces of the code here, so it will be in the upcoming-blocked-by-this-PR issue

File: aclk/agent_cloud_link.c [hunk 14]
```diff
@@ -5,20 +5,30 @@
 
 // Read from the config file -- new section [agent_cloud_link]
 // Defaults are supplied
-int aclk_recv_maximum = 0; // default 20
-int aclk_send_maximum = 0; // default 20
 
-int aclk_port = 0;          // default 1883
-char *aclk_hostname = NULL; //default localhost
+int aclk_port = ACLK_DEFAULT_PORT;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: Will handle all static / variables and function scope in another issue

File: aclk/agent_cloud_link.c [hunk 15]
```diff
@@ -116,14 +103,35 @@ pthread_mutex_t query_lock_wait = PTHREAD_MUTEX_INITIALIZER;
 #define QUERY_THREAD_WAKEUP pthread_cond_signal(&query_cond_wait)
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: Will handle in another issue

File: aclk/mqtt.c [hunk 16]
```diff
@@ -7,34 +7,33 @@
 
 void (*_on_connect)(void *ptr) = NULL;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: Will handle it in the issue with the disconnection detection

File: aclk/mqtt.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: Lets cleanup that in another issue.

File: aclk/mqtt.c [hunk 17]
```diff
@@ -416,7 +391,7 @@ int _link_subscribe(char  *topic, int qos)
  *
  */
 
-int _link_send_message(char *topic, char *message)
+int _link_send_message(char *topic, char *message, int *mid)
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: no special reason  -- if it hurts the eyes (it can) I will rename (beautify) them, in an incoming issue/cleanup

File: aclk/mqtt.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: Will handle in the next issue

File: aclk/mqtt.c [hunk 18]
```diff
@@ -113,10 +79,20 @@ int lws_wss_client_initialized = 0;
 int mqtt_over_websockets = 1;
 struct aclk_lws_wss_engine_instance *lws_engine_instance = NULL;
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: Will handle in the next issue

File: aclk/mqtt.c [hunk 19]
```diff
@@ -432,10 +407,11 @@ int _link_send_message(char *topic, char *message)
     //if (unlikely(rc != MOSQ_ERR_SUCCESS))
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/mqtt.c]: Will remove for now (but keep in mind) for adhoc testing 

Someone needs to take a look at the encoding to make sure I haven't overlooked something ( I will surely do in the next issue in parallel as well)

**Comment by stelfrag** [comment]
> question for the future: when malformed request sent it is logged properly in error log. Should we also inform cloud that we didn't understand?

Good point. It shouldn't happen but it should be logged as it means a lot of things might have gone wrong e.g the versioning is broken and we get stuff we can't handle.

File: aclk/agent_cloud_link.h [hunk 20]
```diff
@@ -6,20 +6,21 @@
 #include "../daemon/common.h"
 #include "mqtt.h"
 
+#define ACLK_VERSION 1
+#define ACLK_THREAD_NAME "ACLK_Query"
 #define ACLK_JSON_IN_MSGID "msg-id"
 #define ACLK_JSON_IN_TYPE "type"
 #define ACLK_JSON_IN_VERSION "version"
 #define ACLK_JSON_IN_TOPIC "callback-topic"
 #define ACLK_JSON_IN_URL "payload"
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.h]: Will do in a next issue

File: aclk/agent_cloud_link.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (stelfrag) [reply aclk/agent_cloud_link.c]: I already put atoi -- but good point :+1:

File: aclk/mqtt.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (underhood) [reply aclk/mqtt.c]: Hi if so we should turn it into enum MQTT_TRANSPORT_TYPE to support later MQTT over HTTPS fallback?

[the remark this answers is not in the document]
Comment (amoss) [reply aclk/mqtt.c]: The HTTPS fall-back option has not been designed or groomed yet. We do not know how the transport layers in it will function. Remove the parts of the code that do not match our use-cases. When we know what the new use-cases are then we will work out what to add.
